sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -850,10 +863,88 @@ std::unique_ptr<CurlSession::CurlConnection> CurlSession::GetCurlConnection(Requ
 
 // Move the connection back to the connection pool. Push it to the front so it becomes the first
 // connection to be picked next time some one ask for a connection to the pool (LIFO)
-void CurlSession::MoveConnectionBackToPool(std::unique_ptr<CurlSession::CurlConnection> connection)
+void CurlConnectionPool::MoveConnectionBackToPool(std::unique_ptr<CurlConnection> connection)
 {
   // Lock mutex to access connection pool. mutex is unlock as soon as lock is out of scope
-  std::lock_guard<std::mutex> lock(s_connectionPoolMutex);
-  auto& hostPool = s_connectionPoolIndex[connection->GetHost()];
+  std::lock_guard<std::mutex> lock(CurlConnectionPool::s_connectionPoolMutex);
+  auto& hostPool = CurlConnectionPool::s_connectionPoolIndex[connection->GetHost()];
+  // update the time when connection was moved back to pool
+  connection->updateLastUsageTime();
   hostPool.push_front(std::move(connection));
+  CurlConnectionPool::s_connectionCounter += 1;
+  // Check if there's no cleaner running and started
+  if (!CurlConnectionPool::s_isCleanConnectionsRunning)
+  {
+    CurlConnectionPool::s_isCleanConnectionsRunning = true;
+    CurlConnectionPool::CleanUp();
+  }
+}
+
+// spawn a thread for cleaning old connections.
+// Thread will keep running while there are at least one connection in the pool
+void CurlConnectionPool::CleanUp()
+{
+  std::thread backgroundCleanerThread([]() {
+    for (;;)
```

Comment:
This is an infinite loop. How does it stop when the program exits?

Reply:
This might crash the whole program

Reply:
it gets stop as soon as there are no connections.  If the clean routine removes all connections, the function returns.

```cpp
if (CurlConnectionPool::s_connectionCounter == 0)
        {
          // stop the cleaner since there are no connections
          CurlConnectionPool::s_isCleanConnectionsRunning = false;
          return;
        }
```

this way, is the cleaner removes all connections, it would wait and if in the next check there are no connections it returns and thread finishes

Reply:
but still, you didn't `join` the thread